clang/lib/3C/RewriteUtils.cpp

```diff
@@ -598,7 +597,20 @@ void RewriteConsumer::emitRootCauseDiagnostics(ASTContext &Context) {
           // SL is invalid when the File is not in the current translation unit.
           if (SL.isValid()) {
             EmittedDiagnostics.insert(PSL);
-            DE.Report(SL, ID) << PtrCount << WReason.second.getWildPtrReason();
+            DE.Report(SL, ID) << PtrCount << WReason.second.getReason();
+          }
+          for (auto &Note : WReason.second.additionalNotes()) {
+            PersistentSourceLoc NPSL = Note.Location;
+            llvm::ErrorOr<const clang::FileEntry *> NFile =
+                SM.getFileManager().getFile(NPSL.getFileName());
+            if (!NFile.getError()) {
```

Comment:
Once I started thinking about multiple files, I thought of another potential problem: Can it happen that a conflict is detected at a location in a header file that is shared by translation units A and B, but the additional notes are in a source file that is only in translation unit A?  If so, it looks like if 3C processes B first (based on the order on the `3c` command line, I think), this code will emit the main root cause diagnostic in the header file and not the notes because their locations are not valid in translation unit A and this condition will be false.  Then the main diagnostic PSL will be added to `EmittedDiagnostics`, so when this code runs on translation unit A, it won't emit any more diagnostics, and we end up never seeing the notes.  It would be better if we could somehow favor a translation unit that includes the notes to ensure that the notes always get emitted.  But if one note is only in translation unit A and the other is only in translation unit B, then we are stuck unless we can somehow get both files into the same SourceManager.  If this problem is uncommon, maybe it makes sense to just add a TODO.  (What is our experience in our porting work so far?)

Reply:
I think you're right that this is a big problem (for code architecture) and may need a new analysis pass for reporting that transcends TUs, just as our info gathering transcends TUs. Currently, the conflicting types must be adjacent in the flow graph, which makes it unlikely that they are in different TUs.

Since the general solution requires architectural changes, a TODO here would probably be ineffective. Would you be willing to open an issue describing the problem and our future needs?

Reply:
Oh, I wasn't paying careful enough attention and I still thought this PR was reporting the reasons for the two edges at the ends of the path (incident on ConstAtoms) as we originally discussed in a meeting.  Now I see your statements that it reports the two edges incident on the VarAtom where the conflict is detected.  You are right that those are unlikely to be in different TUs.

Since you already plan to leave #680 open after this PR to consider further changes to the conflict reporting (e.g., both ends), I think it makes more sense to just add a comment about the multi-TU problem to #680: https://github.com/correctcomputation/checkedc-clang/issues/680#issuecomment-928154627 .  Will you cite this link in the code now?